src/audio/asrc/asrc.c

```diff
@@ -0,0 +1,615 @@
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+	struct sof_ipc_comp_config *config = COMP_GET_CONFIG(dev);
+	struct comp_buffer *sinkb;
+	struct comp_buffer *sourceb;
+	uint32_t source_period_bytes;
+	uint32_t sink_period_bytes;
+	int ret;
+	int frame_bytes;
+	int sample_bytes;
+	int i;
+
+	trace_asrc("asrc_prepare()");
+
+	ret = comp_set_state(dev, COMP_TRIGGER_PREPARE);
+	if (ret < 0)
+		return ret;
+
+	if (ret == COMP_STATUS_STATE_ALREADY_SET)
+		return PPL_STATUS_PATH_STOP;
+
+	/* SRC supports S16_LE, S24_4LE and S32_LE formats */
+	switch (dev->params.frame_fmt) {
+	case SOF_IPC_FRAME_S16_LE:
+		cd->asrc_func = src_copy_s16;
+		break;
+	case SOF_IPC_FRAME_S24_4LE:
+		cd->data_shift = 8;
+		cd->asrc_func = src_copy_s32;
+		break;
+	case SOF_IPC_FRAME_S32_LE:
+		cd->data_shift = 0;
+		cd->asrc_func = src_copy_s32;
+		break;
+	default:
+		trace_asrc_error("asrc_prepare() error: invalid frame format");
+		return -EINVAL;
+	}
+
+	/* SRC component will only ever have 1 source and 1 sink buffer */
+	sourceb = list_first_item(&dev->bsource_list,
+				  struct comp_buffer, sink_list);
+	sinkb = list_first_item(&dev->bsink_list,
+				struct comp_buffer, source_list);
+
+	/* get source data format and period bytes */
+	cd->source_format = comp_frame_fmt(sourceb->source);
+	source_period_bytes = comp_period_bytes(sourceb->source,
+						cd->source_frames);
+
+	/* get sink data format and period bytes */
+	cd->sink_format = comp_frame_fmt(sinkb->sink);
+	sink_period_bytes = comp_period_bytes(sinkb->sink, cd->sink_frames);
+
+	/*
+	 * Allocate input and output data buffer
+	 */
+	frame_bytes = comp_frame_bytes(sourceb->source);
+	cd->buf_size = (cd->source_frames_max + cd->sink_frames_max) *
+		frame_bytes;
+
+	cd->buf = rzalloc(RZONE_RUNTIME, SOF_MEM_CAPS_RAM, cd->buf_size);
+	if (!cd->buf) {
+		cd->buf_size = 0;
+		trace_asrc_error("asrc_prepare() error: rzalloc fail for "
+				 "size %d", cd->buf_size);
+		ret = -ENOMEM;
+		goto err;
+	}
+
+	sample_bytes = frame_bytes / dev->params.channels;
+	for (i = 0; i < dev->params.channels; i++) {
+		cd->ibuf[i] = cd->buf + i * sample_bytes;
+		cd->obuf[i] = cd->ibuf[i] + cd->source_frames_max * frame_bytes;
+	}
+
+	/*
+	 * Get required size and allocate memory for ASRC
+	 */
+	ret = get_required_size_bytes(&cd->src_size, dev->params.channels,
+				      sample_bytes * 8);
+	if (ret) {
+		trace_asrc_error("asrc_prepare() error: "
+				 "get_required_size_bytes");
+		goto err;
+	}
+
+	cd->src_obj = rzalloc(RZONE_RUNTIME, SOF_MEM_CAPS_RAM, cd->src_size);
+	if (!cd->src_obj) {
+		cd->src_size = 0;
+		trace_asrc_error("asrc_prepare() error: rzalloc fail for "
+				 "size %d", cd->src_size);
+		ret = -ENOMEM;
+		goto err;
+	}
+
+	/*
+	 * Initialize ASRC
+	 */
+	ret = initialise_asrc(cd->src_obj, dev->params.channels,
+			      cd->source_rate, cd->sink_rate,
+			      ASRC_IOF_INTERLEAVED, ASRC_IOF_INTERLEAVED,
+			      ASRC_BM_LINEAR, cd->frames, sample_bytes * 8,
+			      ASRC_CM_FEEDBACK, cd->mode);
```

Comment:
I'd like to keep ASRC_BM_xx choice private because it's how I integrate the algorithm as a SOF component. I would have preferred ASRC_BM_CIRCULAR but it currently is only for linear - circular interface and not circular - circular as needed in SOF. After the drift tracking works I will optimize away the cd->buf that currently is used as temporary storage for circular - linear - linear - circular data.

The same goes for ASRC_CM_FEEDBACK. For SOF pipelines currently it's not feasible to fix any processing frames count . To avoid xruns we can process frames .. frames_max. The scheduler today can't handle components those can process only fixed period of samples.

Reply:
Sounds like OK so I'm marking this as resolved.